### PR TITLE
Minting and burning fee

### LINKED ISSUE #202: Minting and burning fee
**Overview**

- Within V2, there will need to be a fee charged both the action of minting and burning. In other words, users will be charged a fee at the time of minting tokens, and also at the time of burning tokens.
- The fee will be a percentage. e.g. 1%. It should be a configurable parameter. The fee should be able to be changed by the contract owner after the time of deployment.
- The fee rate for minting will not necessarily be the same as the fee rate for burning.
- The proceeds of the fee will be added to the pool at which the user was minting into (or burning from).

---

**Minting Fee Logic**

- Hypothetically, let's say the mint fee rate is 1%.
- A user, Bob, commits to minting 100 USDC.
- At the time of committing to mint, 1 USDC will be taken away from the user, Bob. It is 1 USDC because it's 100 USDC * 1%.
- Bob's commit is now only 99 USDC. This will proceed as normal, and will mint tokens at the fair price.

Dev notes:
- In the above example, The user commits 100 Usdc, all 100 are transferred to the pool, but the amount of tokens stored as part of the commit would be 99, so when they go to claim, they will only get that 99 usdc worth of tokens.

---

**Burning Fee Logic**

- Hypothetically, let's say the burn fee rate is 2%.
- A user, Bob, has commited to burn his 200 Long Pool tokens.
- The 200 long pool tokens will sit in the shadow pool until the wait time is up.
- At the rebalancing event in which the commit will be executed, the SC should calculate the value of the 200 long pool tokens (e.g. if price is $2, then the overall value is $400).
- From this, the Smart contract takes $8 ($400 * 2%) from the user, and adds it back into the pool from which it was taken. All the users pool tokens are destroyed.
- 98% of the value is sent to the user.

---

### Extra notes

- Burning fees will require storing the fee rate at in a given update interval (since fee rate can change at any time) and using that to calculate how much quote token a user is entitled to during `updateAggregateBalance`

---

### Completion criteria (added by devs)
Where `n` is the fee rate as a percentage (`n == 1` means fee is 1%)
- When a user mints, `100 - n%` of the settlement tokens they minted with is stored in the user's commitment data.
- When a user mints then claims after update interval, the pool should still be in ownership of `n%` of the original commitment amount of settlement tokens.
- When a user mints, `n%` of the settlement tokens they minted with should be added to the balance of the side to which they are minting.
    - e.g. if they mint $100 to the LONG side, with a 1% fee, directly after minting, $1 should be added to `LeveragedPool::longBalance`
- When a user burns, `100 - n%` of the settlement tokens they would normally be entitled to are in the user's aggregate balance after the update interval passes.
- When a user burns, upkeep happens, then they update their balance (usually through claiming), n% of the settlement tokens they gained should be kept in the pool and that amount added to the correct side of the pool (either `longBalance` or `shortBalance`)
- The same as above, but with `100 - n%` of the settlement tokens deposited to the user's wallet after claiming.
- If `n == 0`, then the user should get the exact amount they would be entitled to before the changes of this PR were implemented

### PR DESCRIPTION
**Pull request by mynameuhh**
- Closes #202 

# Note
The balance invariant that has always existed continues to hold after this change.

# Motivation
We want mint and burn fee

# Changes
**PoolCommitter**
- Add `mintingFee` and `burningFee` for storing minting fee and burning fee as bytes16 ABDKQuad format
- On a mint commitment, transfer the whole amount to pool, take out a percentage from the stored commit `amount`, and add the amount taken to the `longBalance` or `shortBalance` for the side being minted to.
- On a burn commit, store the same data as before (amount of pool tokens burnt)
- When the user next gets their balance updated, using the burn fee at that update interval, take out a percentage of the settlement tokens to which they are entitled. Give the user the rest, and add that small amount to the `longBalance` or `shortBalance` for the side being burned from.

- In `executeCommitments`
    - I further generalised it to remove duplicate code.
    - Store the burn fee for this update interval.

- During updating a user's balance, they receive a bit less settlement tokens based off their burns, due to the burn fee.
- This bit that is taken off is referred to as `_addedLongBalance` or `_addedShortBalance` depending on which side's balance this amount will be added to.

**PoolFactory**
- Now has `mintingFee` and `burningFee` which is used in the exact same way as `fee`.

**PoolSwapLibrary**
In `getUpdatedAggregateBalance`
- Takes `burnFee%` from any settlement tokens earned by burning, and returns it as `_addedShortBalance` or `_addedLongBalance` depending on the side.

### CODE REVIEW

File: contracts/implementation/PoolFactory.sol [hunk 1]
```diff
@@ -109,7 +112,7 @@ contract PoolFactory is IPoolFactory, Ownable {
         PoolCommitter poolCommitter = PoolCommitter(
             Clones.cloneDeterministic(poolCommitterBaseAddress, uniquePoolHash)
         );
-        poolCommitter.initialize(address(this));
+        poolCommitter.initialize(address(this), mintingFee, burningFee);
```
Comment (Paul4912) [inline contracts/implementation/PoolFactory.sol:115]: don't we need to divide by each update interval in the year? this fee seems to be used as is in each commit.

Comment (Paul4912) [reply contracts/implementation/PoolFactory.sol:115]: better way to phrase the question is minting and burning fee meant to be a flat fee of every mint and burn on top of yearly maintenance fee? 

OR

is the mint and burn fee meant to be an annual fee that we want to divide up into every update interval and extract over the year just like with maintenance fee?

File: contracts/implementation/PoolSwapLibrary.sol [hunk 2]
```diff
@@ -417,16 +418,18 @@ library PoolSwapLibrary {
      */
     function getUpdatedAggregateBalance(UpdateData calldata data)
         external
-        pure
+        view
         returns (
             uint256 _newLongTokens,
             uint256 _newShortTokens,
+            uint256 _addedLongBalance,
```
Comment (Paul4912) [inline contracts/implementation/PoolSwapLibrary.sol:425]: From my understanding this is the amount that is extracted from the burn as a burning fee. addedLongBalance sounds vague why not _longBurnFee and _shortBurnFee

**Comment by Paul4912** [comment]
So when the mint or burn fee is taken and added to the pool balance it essentially donates the feel to rest of the people already holding the long or short token correct? What was motivation behind this change financially, to make holding a little better compared to actively trading?

File: contracts/implementation/PoolFactory.sol (hunk 1 above)
Comment (mynameuhh) [reply contracts/implementation/PoolFactory.sol:115]: The spec given to me explicitly said per commit, so the former :shell:

File: contracts/implementation/PoolSwapLibrary.sol (hunk 2 above)
Comment (mynameuhh) [reply contracts/implementation/PoolSwapLibrary.sol:425]: Yep good idea. I suck at variable naming

**Comment by mynameuhh** [comment]
@Paul4912 I think the motivation was as a mitigation for volatility decay (on top of other mitigations). So yeah basically to make holding a more viable option

**Review by Paul4912 — APPROVED** [review]

**Review by sporejack — CHANGES_REQUESTED** [review]

File: contracts/implementation/PoolCommitter.sol [hunk 3]
```diff
@@ -15,13 +15,18 @@ contract PoolCommitter is IPoolCommitter, Initializable {
     uint128 public constant LONG_INDEX = 0;
     uint128 public constant SHORT_INDEX = 1;
 
+    // The amount that is extracted from each mint and burn, being left in the pool. Given as the decimal * 10 ^ 18. For example, 60% fee is 0.6 * 10 ^ 18
+    bytes16 mintingFee;
+    bytes16 burningFee;
```
[the remark this answers is not in the document]
Comment (mynameuhh) [reply contracts/implementation/PoolCommitter.sol:20]: They are set in `initialize` so I'm not sure if that is needed

File: contracts/implementation/PoolFactory.sol [hunk 4]
```diff
@@ -29,8 +29,11 @@ contract PoolFactory is IPoolFactory, Ownable {
 
     // Contract address to receive protocol fees
     address public feeReceiver;
-    // Default fee; Fee value as a decimal multiplied by 10^18. For example, 0.5% is represented as 0.5 * 10^18
+    // Default fee, annualised; Fee value as a decimal multiplied by 10^18. For example, 50% is represented as 0.5 * 10^18
     uint256 public fee;
+    // The fee taken for each mint and burn. Fee value as a decimal multiplied by 10^18. For example, 50% is represented as 0.5 * 10^18
+    uint256 public mintingFee;
+    uint256 public burningFee;
```
[the remark this answers is not in the document]
Comment (mynameuhh) [reply contracts/implementation/PoolFactory.sol:36]: I followed the same pattern as with the normal `fee`. What do you think should happen here?

**Review by sporejack — APPROVED** [review]

**Pull request merged**